docs/proposals/PROPOSAL_F.md

```diff
@@ -0,0 +1,374 @@
+            "org.opencontainers.signatures.index":"dev.cosignproject.cosign/signature.e59879c,dev.cosignproject.cosign/signature.de25bf6,dev.sigstore.cosign/bundle.37d34ff,dev.sigstore.cosign/certificate.e8adf60,dev.sigstore.cosign/chain.b1c64a3,dev.sigstore.cosign/bundle.6436749,dev.sigstore.cosign/certificate.776a3f1,dev.sigstore.cosign/chain.8adf60e"
+         }
+      },
+      {
+         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
+         "digest": "sha256:4ff3ca91275773af45cb4b0834e12b7eb47d1c18f770a0b151381cd227f4c253",
+         "size": 528,
+         "platform": {
+            "architecture": "amd64",
+            "os": "linux"
+         },
+         // signatures added to the linux/amd64
+         "annotations": {
+            "dev.cosignproject.cosign/signature.e59879c": "...",
+            "dev.sigstore.cosign/bundle.6436749": "...",
+            "dev.sigstore.cosign/certificate.776a3f1": "...",
+            "dev.sigstore.cosign/chain.8adf60e": "...",
+            "dev.cosignproject.cosign/signature.de25bf6": "...",
+            "dev.sigstore.cosign/bundle.37d34ff": "...",
+            "dev.sigstore.cosign/certificate.e8adf60": "...",
+            "dev.sigstore.cosign/chain.b1c64a3": "...",
+            "org.opencontainers.signatures.index":"dev.cosignproject.cosign/signature.e59879c,dev.cosignproject.cosign/signature.de25bf6,dev.sigstore.cosign/bundle.37d34ff,dev.sigstore.cosign/certificate.e8adf60,dev.sigstore.cosign/chain.b1c64a3,dev.sigstore.cosign/bundle.6436749,dev.sigstore.cosign/certificate.776a3f1,dev.sigstore.cosign/chain.8adf60e"
+         }
+      },
+      // artifacts referencing the image manifests
+      {
+         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
+         "digest": "sha256:5b0044a1244...", // arm64 SBOM
+         "size": 1024,
+         "annotations": {
+            "org.opencontainers.reference.type": "sbom",
+            "org.opencontainers.reference.digest": "sha256:c3c58223e2af75154c4a7852d6924b4cc51a00c821553bbd9b3319481131b2e0",
+            "org.opencontainers.reference.description": "Software Bill of Materials of alpine:3.17 linux/arm64 image"
+         }
+      },
+      {
+         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
+         "digest": "sha256:5b0044a44d...", // amd64 SBOM
+         "size": 1024,
+         "annotations": {
+            "org.opencontainers.reference.type": "sbom",
+            "org.opencontainers.reference.digest": "sha256:4ff3ca91275773af45cb4b0834e12b7eb47d1c18f770a0b151381cd227f4c253",
+            "org.opencontainers.reference.description": "Software Bill of Materials of alpine:3.17 linux/amd64 image"
+         }
+      }
+   ]
+}
+```
+
+### Registry HTTP API
+
+> The distribution-spec **SHOULD** solve the race condition occurring when [multiple parties modify an OCI Index in parallel](https://github.com/opencontainers/distribution-spec/pull/251).
```

Comment:
For existing registries that don't have this feature yet, are we accepting the race condition?

Reply:
I think that we need to. Depending on the time horizon for fixing this, we could provide guidance for how clients can work around the race.

I would worry that the guidance would become the standard though 😬

Reply:
We could say that, if the response doesn't include an `ETag`, that the client `SHOULD` re-pull the manifest and check that the contents are what it expects, and if it doesn't, it `SHOULD` go through the pull-update-push dance again, until the update lands racelessly. Or, client's `MAY` consider the case an error and make users re-push until the race doesn't happen, possibly losing information along the way.

This seems like both a.) enough to ensure the race doesn't practically happen, and b.) a slow and annoying enough protocol that it still encourages registries to support ETags so it isn't necessary. If a client wants to assume the registry will support ETags they can, and just fail loudly if it doesn't.

This has the possible negative side effect of pushing the updated manifest multiple times, each time it may lose the race and need to retry. Again, this is more reason a registry should prefer not to need this, and just support ETags.

Ultimately there's not a lot we can do to make the protocol bullet-proof, except just promote ETags everywhere.

Reply:
Depending on how long it takes to push various blobs, the client could have cached a manifests for a long while (in computer time) and overwrite a manifest that was pushed and verified by another client. In other words, there are lots of races, and we'll probably need to spend some effort specifying the client requirements to minimize (but not prevent) these races.

We also need to factor in that some tooling may see an artifact get pushed and attempt to intentionally delete that artifact, and we'd need a way to differentiate that from a race condition.